src/devices/CharacterLcd/Hd44780.cs

```diff
@@ -68,181 +77,312 @@ public class Hd44780 : Hd44780Base, IDisposable
+                default:
+                    // We don't support other rows, users can derive for odd cases.
+                    // (Three row LCDs exist, but aren't common.)
+                    throw new ArgumentOutOfRangeException(nameof(rows));
             }
+
+            return rowOffsets;
         }
 
         /// <summary>
-        /// Enable/disable the backlight. (Will always return false if no backlight pin was provided.)
+        /// Wait for the device to not be busy.
         /// </summary>
-        public override bool BacklightOn
+        /// <param name="microseconds">Time to wait if checking busy state isn't possible/practical.</param>
+        protected void WaitForNotBusy(int microseconds)
         {
-            get
-            {
-                return _backlight != -1 && _controller.Read(_backlight) == PinValue.High;
-            }
-            set
-            {
-                if (_backlight != -1)
-                    _controller.Write(_backlight, value ? PinValue.High : PinValue.Low);
-            }
+            _interface.WaitForNotBusy(microseconds);
         }
 
         /// <summary>
-        /// Send a data or command byte to the controller.
+        /// Clears the LCD, returning the cursor to home and unshifting if shifted.
+        /// Will also set to Increment.
         /// </summary>
-        /// <param name="data">True to send data, otherwise sends a command.</param>
-        protected override void Send(byte value, bool data = false)
+        public void Clear()
         {
-            _controller.Write(_rsPin, data ? PinValue.High : PinValue.Low);
+            SendCommand(ClearDisplayCommand);
+            WaitForNotBusy(3000);
+        }
 
-            if (_rwPin != -1)
-            {
-                _controller.Write(_rwPin, PinValue.Low);
-            }
+        /// <summary>
+        /// Moves the cursor to the first line and first column.
+        /// </summary>
+        public void Home()
+        {
+            SendCommand(ReturnHomeCommand);
+            WaitForNotBusy(3000);
+        }
 
-            if (_dataPins.Length == 8)
-            {
-                WriteBits(value, 8);
-            }
-            else
-            {
-                WriteBits((byte)(value >> 4), 4);
-                WriteBits(value, 4);
-            }
+        /// <summary>
+        /// Moves the cursor to an explicit column and row position.
+        /// </summary>
+        /// <param name="left">The column position from left to right starting with 0.</param>
+        /// <param name="top">The row position from the top starting with 0.</param>
+        public void SetCursorPosition(int left, int top)
+        {
+            int rows = _rowOffsets.Length;
+            if (top < 0 || top >= rows)
+                throw new ArgumentOutOfRangeException(nameof(top));
+
+            int newAddress = left + _rowOffsets[top];
+            if (left < 0 || (rows == 1 && newAddress >= 80) || (rows > 1 && newAddress >= 104))
```

Comment:
I think the numbers should have some comment or reference to datasheet

Reply:
The row setup code explains the addressing in detail. I'll add a comment here.